In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [2]:
# Read in CSV file
regional_df = pd.read_csv("Table_CSVs/regional.csv")
regional_df

Unnamed: 0   DOEID    REGIONC            DIVISION state_postal  \
0               0  100001       WEST      Mountain South           NM   
1               1  100002      SOUTH  West South Central           AR   
2               2  100003       WEST      Mountain South           NM   
3               3  100004      SOUTH      South Atlantic           SC   
4               4  100005  NORTHEAST     Middle Atlantic           NJ   
...           ...     ...        ...                 ...          ...   
18491       18491  118492      SOUTH      South Atlantic           MD   
18492       18492  118493  NORTHEAST         New England           ME   
18493       18493  118494      SOUTH  West South Central           TX   
18494       18494  118495      SOUTH      South Atlantic           SC   
18495       18495  118496    MIDWEST  West North Central           MO   

        BA_climate   TOTALBTU  TOTALDOL  
0        Mixed-Dry  144647.71   2656.89  
1      Mixed-Humid   28034.61    975.00  
2        Mixed-Dry   30749.71    522.65  
3      Mixed-Humid   86765.19   2061.77  
4      Mixed-Humid   59126.93   1463.04  
...            ...        ...       ...  
18491  Mixed-Humid   49930.49   1098.51  
18492    Very-Cold  222186.04   3613.44  
18493    Hot-Humid   51593.72   1428.31  
18494    Hot-Humid   63555.21   2224.94  
18495  Mixed-Humid   67621.81   1957.28  

[18496 rows x 8 columns]

In [3]:
# Create DataFrame with regional information
regional_df = regional_df[['REGIONC', 'DIVISION', 'state_postal', 'BA_climate', 'TOTALBTU']]
regional_df.head()

REGIONC            DIVISION state_postal   BA_climate   TOTALBTU
0       WEST      Mountain South           NM    Mixed-Dry  144647.71
1      SOUTH  West South Central           AR  Mixed-Humid   28034.61
2       WEST      Mountain South           NM    Mixed-Dry   30749.71
3      SOUTH      South Atlantic           SC  Mixed-Humid   86765.19
4  NORTHEAST     Middle Atlantic           NJ  Mixed-Humid   59126.93

In [4]:
#Checking nulls
regional_df.isnull().sum()

REGIONC         0
DIVISION        0
state_postal    0
BA_climate      0
TOTALBTU        0
dtype: int64

In [5]:
#Separate the features `X` from the target `y`
y = regional_df['TOTALBTU']
X = regional_df.drop(columns='TOTALBTU')

In [6]:
# Preview the features data
X.head()

REGIONC            DIVISION state_postal   BA_climate
0       WEST      Mountain South           NM    Mixed-Dry
1      SOUTH  West South Central           AR  Mixed-Humid
2       WEST      Mountain South           NM    Mixed-Dry
3      SOUTH      South Atlantic           SC  Mixed-Humid
4  NORTHEAST     Middle Atlantic           NJ  Mixed-Humid

In [7]:
# Preview the first five entries for the target variable
y[:5]

0    144647.71
1     28034.61
2     30749.71
3     86765.19
4     59126.93
Name: TOTALBTU, dtype: float64

In [8]:
# Encode the categorical variables using get_dummies
X = pd.get_dummies(X)

# Review the features data
X.head()

REGIONC_MIDWEST  REGIONC_NORTHEAST  REGIONC_SOUTH  REGIONC_WEST  \
0            False              False          False          True   
1            False              False           True         False   
2            False              False          False          True   
3            False              False           True         False   
4            False               True          False         False   

   DIVISION_East North Central  DIVISION_East South Central  \
0                        False                        False   
1                        False                        False   
2                        False                        False   
3                        False                        False   
4                        False                        False   

   DIVISION_Middle Atlantic  DIVISION_Mountain North  DIVISION_Mountain South  \
0                     False                    False                     True   
1                     False                    False                    False   
2                     False                    False                     True   
3                     False                    False                    False   
4                      True                    False                    False   

   DIVISION_New England  ...  state_postal_WV  state_postal_WY  \
0                 False  ...            False            False   
1                 False  ...            False            False   
2                 False  ...            False            False   
3                 False  ...            False            False   
4                 False  ...            False            False   

   BA_climate_Cold  BA_climate_Hot-Dry  BA_climate_Hot-Humid  \
0            False               False                 False   
1            False               False                 False   
2            False               False                 False   
3            False               False                 False   
4            False               False                 False   

   BA_climate_Marine  BA_climate_Mixed-Dry  BA_climate_Mixed-Humid  \
0              False                  True                   False   
1              False                 False                    True   
2              False                  True                   False   
3              False                 False                    True   
4              False                 False                    True   

   BA_climate_Subarctic  BA_climate_Very-Cold  
0                 False                 False  
1                 False                 False  
2                 False                 False  
3                 False                 False  
4                 False                 False  

[5 rows x 73 columns]

In [9]:
# Split the dataset using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,  random_state=42)

In [10]:
# Create a model with scikit-learn
model = LinearRegression()

In [11]:
# Fit the data into the model / train the model
model.fit(X_train, y_train)

LinearRegression()

In [12]:
# Make predictions using the X set
predictions = model.predict(X_test)

In [13]:
#CHECKING THE TRAINING RESULTS
y_train_pred = model.predict(X_train)

r2_training = r2_score(y_train, y_train_pred)
print("R-squared for Training Data:", r2_training)

R-squared for Training Data: 0.12483666166541985


In [14]:
# Compute metrics for the linear regression model: score, r2, mse, rmse, std
score = model.score(X, y, sample_weight=None)
r2 = r2_score(y_test, predictions)
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
std = np.std(y)

# Print relevant metrics.
print(f"The score is {score}.")
print(f"The r2 is {r2}.")
print(f"The mean squared error is {mse}.")
print(f"The root mean squared error is {rmse}.")
print(f"The standard deviation is {std}.")

The score is 0.12131530423001635.
The r2 is 0.1097423932846261.
The mean squared error is 2355471249.4983034.
The root mean squared error is 48533.19739619783.
The standard deviation is 53204.5566692713.


In [15]:
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)

Coefficients: [ 3.90092358e+16  6.42761413e+17  3.00074281e+17 -1.31817086e+17
  3.39274837e+17  7.58148228e+16 -3.80295278e+17  4.11829939e+17
  3.77252534e+17 -4.28439935e+17  4.69385144e+17  2.80482347e+16
  2.33632269e+17 -1.32457208e+17 -2.11032138e+16 -5.94242604e+16
  1.48847770e+17  7.10293961e+16 -2.11032138e+16  3.64519911e+16
  1.02143366e+17 -1.16576723e+16 -1.16576723e+16 -1.16576723e+16
 -1.16576723e+16 -2.11032138e+16  4.38233392e+16  3.64519911e+16
 -6.18192293e+16 -6.18192293e+16  4.38233392e+16 -5.94242604e+16
  1.48847770e+17  1.02143366e+17 -1.16576723e+16  1.02143366e+17
 -6.18192293e+16  4.38233392e+16  4.38233392e+16 -5.94242604e+16
  3.64519911e+16 -1.16576723e+16  4.38233392e+16  4.38233392e+16
  1.02143366e+17  5.39987081e+16  7.10293961e+16  7.10293961e+16
  5.39987081e+16 -6.18192293e+16  1.48847770e+17 -2.11032138e+16
  5.39987081e+16  1.02143366e+17 -1.16576723e+16  4.38233392e+16
 -5.94242604e+16  1.48847770e+17  3.64519911e+16 -1.16576723e+16
  1.0214336

Random Forest

In [16]:
# Create a random forest classifier
RF_model= RandomForestRegressor(n_estimators=100, random_state=42)

In [17]:
# Fitting the model
RF_model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [18]:
#Making predictions
y_pred = RF_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)


Mean Squared Error: 2364802816.8932896
R-squared: 0.10621549867373536


In [21]:
# Perform 5-fold cross-validation

from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(RF_model, X, y, cv=5, scoring='r2')

# Print the cross-validation scores
print("Cross-Validation R-squared Scores:", cv_scores)
print("Mean R-squared Score:", cv_scores.mean())

Cross-Validation R-squared Scores: [0.11372301 0.13425983 0.08887345 0.12044533 0.1208428 ]
Mean R-squared Score: 0.11562888491132732


 the R-squared scores range from approximately 0.089 to 0.134 across the five folds, with a mean R-squared score of approximately 0.116. This suggests that the model's overall ability to explain the variance in the target variable is limited, and there may be room for improvement in its performance.

In [22]:
# Checking Feature importance
feature_importance = pd.Series(RF_model.feature_importances_, index=X.columns)
print("Feature Importance:")
# We can sort the features by their importance
sorted(zip(RF_model.feature_importances_, X.columns), reverse=True)

Feature Importance:


[(0.3592202624922513, 'BA_climate_Cold'),
 (0.13729014200474143, 'state_postal_AK'),
 (0.11197371127419131, 'BA_climate_Mixed-Humid'),
 (0.06424360287504426, 'BA_climate_Very-Cold'),
 (0.037925744033753137, 'state_postal_HI'),
 (0.036384783726627686, 'REGIONC_SOUTH'),
 (0.03356074594717283, 'state_postal_NJ'),
 (0.022739978852014214, 'state_postal_CA'),
 (0.01897627028790957, 'DIVISION_South Atlantic'),
 (0.016539956508360425, 'REGIONC_MIDWEST'),
 (0.015783153067096725, 'state_postal_DC'),
 (0.011579575861862451, 'DIVISION_Pacific'),
 (0.009794390486733537, 'state_postal_NY'),
 (0.009582933587806063, 'state_postal_NC'),
 (0.008247942717753018, 'DIVISION_Mountain South'),
 (0.006761614069735749, 'state_postal_MI'),
 (0.006035893294178419, 'state_postal_NV'),
 (0.005803688053395209, 'state_postal_SC'),
 (0.004875309409361774, 'DIVISION_East North Central'),
 (0.003934521839940046, 'DIVISION_West North Central'),
 (0.0039027499196768838, 'DIVISION_Mountain North'),
 (0.0038730316692716267

Optimizing Random Forest

In [23]:
#initialize model
optimized_rf = RandomForestRegressor()

In [24]:
#define hyperparameters tuning
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['auto', 'sqrt', 'log2']
}

In [25]:
# Instantiate the GridSearchCV object
#This object will systematically test different combinations of hyperparameters using cross-validation to identify the optimal configuration for the model.
grid_search = GridSearchCV(
    estimator=optimized_rf, 
    param_grid=param_grid, 
    cv=5, # Number of cross-validation folds (popular choice)
    scoring='neg_mean_squared_error', # Scoring method for evaluation
    verbose=2 # Verbosity level
    )

# Fit the model
grid_search.fit(X_train, y_train)

Fitting 5 folds for each of 324 candidates, totalling 1620 fits
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=

c:\ProgramData\anaconda3\envs\dev\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
540 fits failed out of a total of 1620.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
540 fits failed with the following error:
Traceback (most recent call last):
  File "c:\ProgramData\anaconda3\envs\dev\lib\site-packages\sklearn\model_selection\_validation.py", line 732, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\ProgramData\anaconda3\envs\dev\lib\site-packages\sklearn\base.py", line 1144, in wrapper
    estimator._validate_params()
  File "c:\ProgramData\anaconda3\envs\dev\lib\site-packages\sklearn\base.py", line 637, in _validate_params
    validate_parameter_constraints(
  File "c:\

GridSearchCV(cv=5, estimator=RandomForestRegressor(),
             param_grid={'max_depth': [None, 5, 10, 15],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300]},
             scoring='neg_mean_squared_error', verbose=2)

In [26]:
# Print the best hyperparameters
print('Best hyperparameters:', grid_search.best_params_)
print('Best accuracy score:', grid_search.best_score_)

Best hyperparameters: {'max_depth': 15, 'max_features': 'sqrt', 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 100}
Best accuracy score: -2551712141.416862


Using te best parameters

In [27]:
# Instantiate the RandomForestRegressor model with the best hyperparameters
new_optimized_rf = RandomForestRegressor(
    n_estimators=100,
    max_depth=15,
    min_samples_split=10,
    min_samples_leaf=4,
    max_features='log2'
)

In [28]:
# Fit the optimized model to the training data
new_optimized_rf.fit(X_train, y_train)

RandomForestRegressor(max_depth=15, max_features='log2', min_samples_leaf=4,
                      min_samples_split=10)

In [29]:
# Make predictions using the optimized model
y_pred_new = new_optimized_rf.predict(X_test)

In [30]:
# Evaluate the model's performance using the test data
mse = mean_squared_error(y_test, y_pred_new)
r2 = r2_score(y_test, y_pred_new)
print("Mean Squared Error on Test Data:", mse)
print("R-squared:", r2)

Mean Squared Error on Test Data: 2355316178.509275
R-squared: 0.10980100284209482
